site/content/blog/2024-12-24-use-zap-with-flagger-in-kubernetes/index.md

```diff
@@ -0,0 +1,575 @@
+
+1. Add the Flagger helm repository
+    ```
+    helm repo add flagger https://flagger.app
+    ```
+2. The Flagger Load Tester will need permission to create Jobs in the cluster, so we'll install it with the option to set up a role for itself.
+    ```
+    helm install flagger-loadtester flagger/loadtester --namespace zap-demo --set rbac.create=true --set rbac.rules[0].apiGroups[0]="batch" --set "rbac.rules[0].resources={jobs,jobs/log,cronjobs,cronjobs/log}" --set "rbac.rules[0].verbs={get,create,list,watch}"
+    ```
+
+## Deploy Testing Service
+The podinfo container will be used as our application that Flagger will be coordinating for deployment.
+1. Copy the podinfo `Deployment` and `Service` into a `podinfo.yaml` file and run `kubectl apply -f podinfo.yaml`
+    ```
+    apiVersion: apps/v1
+    kind: Deployment
+    metadata:
+      name: podinfo
+      namespace: zap-demo
+    spec:
+      selector:
+        matchLabels:
+          app: podinfo
+      template:
+        metadata:
+          labels:
+            app: podinfo
+        spec:
+          containers:
+          - name: podinfod
+            image: ghcr.io/stefanprodan/podinfo:6.7.0
+            ports:
+            - name: http
+              containerPort: 9898
+              protocol: TCP
+            command:
+            - ./podinfo
+            - --port=9898
+            - --level=info
+    ---
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: podinfo
+      namespace: zap-demo
+    spec:
+      type: ClusterIP
+      selector:
+        app: podinfo
+      ports:
+        - name: http
+          port: 9898
+          protocol: TCP
+          targetPort: http
+    ```
+
+## Deploy ZAP resources
+Flagger will be triggering ZAP, but we need to set up the resources ahead of time.
+
+1. Copy the following `PersistentVolumeClaim`, substituting the `storageClassName` value with an appropriate `StorageClass` on your cluster. Save as `zap-pvc.yaml` and apply with `kubectl apply -f zap-pvc.yaml`. This is where the ZAP reports will be saved. I'm saving mine to Azure Blob Storage, which is where I will be reading the report from.
+    ```
+    apiVersion: v1
+    kind: PersistentVolumeClaim
+    metadata:
+      labels:
+        app.kubernetes.io/name: zap
+      name: zap-pvc
+      namespace: zap-demo
+    spec:
+      accessModes:
+      - ReadWriteOnce
+      resources:
+        requests:
+          storage: 50Mi
+      storageClassName: dynamic-slrs-blob-fuse
+    ```
+2. Create the following `CronJob` as `zap.yaml` and apply with `kubectl apply -f zap.yaml`. We are creating it suspended, so it will never run on the schedule, however we will be using it for Flagger to create the ZAP `Job`.
```

Comment:
Just curious, does it have to be a CronJob? Why not make it a regular (suspended) Job?

Reply:
Good question, everything was setup to be GitOps compatible. In theory, I think a suspended Job that is resumed during the rollout would work, but once GitOps is thrown into the mix, one may have extra steps to ensure that the GitOps tooling doesn't revert the change to the Job.

If trying to use the Job as a template like with the CronJob, AFAIK, the same CLI syntax couldn't be used. I think one would have to retrieve the Job, override the suspend and apply the new job. But then it would be missing some things you get for free when using the CronJob, like Job history getting cleaned up, and deleting of the CronJob then deletes all the child jobs.